### PR TITLE
fix(ingest/tableau): project_path_pattern use in _is_denied_project

### PR DESCRIPTION
**Pull request by sid-acryl**

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py [hunk 1]
```diff
@@ -908,8 +908,11 @@ def _is_allowed_project(self, project: TableauProject) -> bool:
...
-        # Either project name or project path should exist in deny
-        for deny_pattern in self.config.project_pattern.deny:
+        # Either project_pattern or project_path_pattern is set
+        deny_patterns: List[str] = self.config.project_pattern.deny
+        deny_patterns.extend(self.config.project_path_pattern.deny)
+
+        for deny_pattern in deny_patterns:
             # Either name or project path is denied
             if re.match(
                 deny_pattern, project.name, self.config.project_pattern.regex_flags
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:918]: Why are we doing this? We should match the project name against `self.config.project_pattern` and the project path against `self.config.project_path_pattern`, but matching both values against both patterns does not make sense

Are there some backwards compat concerns here? If so, what are they?

**Review by hsheth2 — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py (hunk 1 above)
Comment (sid-acryl) [reply metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:917]: Simplified the code

**Review by hsheth2 — APPROVED** [review]
overall this code still confuses me

why do we have both is_allowed_project and is_denied_project? why isn't one method  sufficient here? also, ideally we'd just use `AllowDenyPattern`'s built-in methods instead of re-implementing a bunch of regex matching logic here

**Comment by sid-acryl** [comment]
> overall this code still confuses me
> 
> why do we have both is_allowed_project and is_denied_project? why isn't one method sufficient here? also, ideally we'd just use `AllowDenyPattern`'s built-in methods instead of re-implementing a bunch of regex matching logic here

@hsheth2 
It took me while to recall why I did it like this. I added a below doc string in the function `_is_denied_project` and also renamed the `AllowDenyPattern._denied` to `AllowDenyPattern.denied` to reuse the regx logic.

        """
        Why use an explicit denial check instead of the `AllowDenyPattern.allowed` method?

        Consider a scenario where a Tableau site contains four projects: A, B, C, and D, with the following hierarchical relationship:

        - **A**
          - **B** (Child of A)
          - **C** (Child of A)
        - **D**

        In this setup:

        - `project_pattern` is configured with `allow: ["A"]` and `deny: ["B"]`.
        - `extract_project_hierarchy` is set to `True`.

        The goal is to extract assets from project A and its children while explicitly denying the child project B.

        If we rely solely on the `project_pattern.allowed()` method, project C's assets will not be ingested.
        This happens because project C is not explicitly included in the `allow` list, nor is it part of the `deny` list.
        However, since `extract_project_hierarchy` is enabled, project C should ideally be included in the ingestion process unless explicitly denied.

        To address this, the function explicitly checks the deny regex to ensure that project C’s assets are ingested if it is not specifically denied in the deny list. This approach ensures that the hierarchy is respected while adhering to the configured allow/deny rules.
        """
        
 I will create a separate PR to add test cases for this scenario.

**Review by hsheth2 — APPROVED** [review]

**Comment by hsheth2** [comment]
@sid-acryl let's create some unit tests for that scenario

**Comment by sid-acryl** [comment]
> @sid-acryl let's create some unit tests for that scenario

@hsheth2 
Added test-case here https://github.com/datahub-project/datahub/pull/12079

**Pull request merged**